data/scripts.js

```diff
@@ -300,165 +263,259 @@ let BattleScripts = {
+			this.hitStepTryHitEvent,
+
+			// 2. check for type immunity (this is step 1 in gens 4-6)
+			this.hitStepTypeImmunity,
+
+			// 3. check for powder immunity (gen 6+ only)
+			this.hitStepPowderImmunity,
+
+			// 4. check for prankster immunity (gen 6+ only)
+			this.hitStepPranksterImmunity,
+
+			// 5. check accuracy
+			this.hitStepAccuracy,
+
+			// 6. break protection effects
+			this.hitStepBreakProtect,
+
+			// 7. steal positive boosts (Spectral Thief)
+			this.hitStepStealBoosts,
+
+			// 8. loop that processes each hit of the move (has its own steps per iteration)
+			this.hitStepMoveHitLoop,
+		];
+		if (this.gen <= 6) {
+			// Swap step 1 with step 2
+			[moveSteps[1], moveSteps[2]] = [moveSteps[2], moveSteps[1]];
+		}
+		if (this.gen === 4) {
+			// Swap step 5 with new step 2 (old step 1)
+			[moveSteps[2], moveSteps[5]] = [moveSteps[5], moveSteps[2]];
+		}
+
+		this.setActiveMove(move, pokemon, targets[0]);
 		move.zBrokeProtect = false;
 
-		let hitResult = this.singleEvent('PrepareHit', move, {}, target, pokemon, move);
+		let hitResult = this.singleEvent('PrepareHit', move, {}, targets[0], pokemon, move);
 		if (!hitResult) {
 			if (hitResult === false) {
 				this.add('-fail', pokemon);
 				this.attrLastMove('[still]');
 			}
 			return false;
 		}
-		this.runEvent('PrepareHit', pokemon, target, move);
+		this.runEvent('PrepareHit', pokemon, targets[0], move);
 
-		if (!this.singleEvent('Try', move, null, pokemon, target, move)) {
+		if (!this.singleEvent('Try', move, null, pokemon, targets[0], move)) {
 			return false;
 		}
 
-		if (move.target === 'all' || move.target === 'foeSide' || move.target === 'allySide' || move.target === 'allyTeam') {
-			if (move.target === 'all') {
-				hitResult = this.runEvent('TryHitField', target, pokemon, move);
-			} else {
-				hitResult = this.runEvent('TryHitSide', target, pokemon, move);
-			}
-			if (!hitResult) {
-				if (hitResult === false) {
-					this.add('-fail', pokemon);
-					this.attrLastMove('[still]');
-				}
-				return false;
+		let atLeastOneFailure;
+		for (const step of moveSteps) {
+			/** @type {(number | boolean | "" | undefined)[] | undefined} */
+			let hitResults = step.call(this, targets, pokemon, move);
+			if (!hitResults) continue;
+			// @ts-ignore
+			targets = targets.filter((val, i) => hitResults[i] || hitResults[i] === 0);
+			atLeastOneFailure = atLeastOneFailure || hitResults.some(val => val === false);
+			if (!targets.length) {
+				// console.log(step.name);
+				break;
 			}
-			return this.moveHit(target, pokemon, move);
 		}
 
-		hitResult = this.runEvent('TryImmunity', target, pokemon, move);
-		if (!hitResult) {
-			if (hitResult !== null) {
+		const moveResult = !!targets.length;
+		if (!moveResult && !atLeastOneFailure) pokemon.moveThisTurnResult = null;
+		if (move.spreadHit) this.attrLastMove('[spread] ' + targets.join(','));
```

Comment:
You forgot to use `hitTargets` (technically an array of slots) instead of `targets`. You did it correctly in Gen 3, too.